blocks/api/categories.js

```diff
@@ -21,10 +24,121 @@ const categories = [
 ];
 
 /**
- * Returns all the block categories
+ * @type {RegExp}
+ * @const
+ *
+ * Category names must be a combination of lower-case letters, numbers, and hypens
+ */
+const categoryNamePattern = /^[a-z0-9-]+$/;
+
+/**
+ * Returns all the block categories.
  *
- * @return {Array} Block categories
+ * @return {Array} categories
  */
 export function getCategories() {
 	return categories;
 }
+
+/**
+ * Register a new block Category.
+ *
+ * @param {Array} category e.g {slug: 'custom', title: __('Custom Blocks')}
+ *
+ * @return {Array} categories
+ */
+export function registerCategory( category ) {
+	if ( ! category ) {
+		console.error(
+			'The block Category must be defined'
+		);
+		return;
+	}
+	if ( ! category.slug ) {
+		console.error(
+			'The block Category slug must be defined'
+		);
+		return;
+	}
+	if ( ! categoryNamePattern.test( category.slug ) ) {
+		console.error(
+			'The block Category slug must not contain characters which are invalid for urls'
+		);
+		return;
+	}
+
+	if ( categories.some( x => x.slug === category.slug ) ) {
+		console.error(
+			'The block Category "' + category.slug + '" is already registered'
+		);
+		return;
+	}
+	if ( ! category.title ) {
+		console.error(
+			'The block Category title must be defined'
+		);
+		return;
+	}
+
+	categories.push( category );
+	return categories;
+}
```

Comment:
oh typing…we either get it in a robust and helpful way or we manually force it in unreliable and costly ways…

(this is not a comment on the code submission, but just me noting a place where having static typing could save us so much effort _and_ provide us so much help)